servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/H2PriorKnowledgeFeatureParityTest.java

```diff
@@ -435,6 +442,126 @@ public void serverHeaderCookieRemovalAndIteration() throws Exception {
+            } else {
+                ReservedBlockingHttpConnection reservedConn = client.reserveConnection(request);
+                try {
+                    reservedConn.request(request);
+                    assertThrows(ClosedChannelException.class, () ->
+                            reservedConn.request(client.get("/").payloadBody("a", textSerializer())));
+                } finally {
+                    assertThrows(IllegalStateException.class, reservedConn::release);
+                }
+            }
+        }
+    }
+
+    @Test
+    public void serverSendsLargerContentLength() throws Exception {
+        assumeTrue(h2PriorKnowledge); // http/1.x will timeout waiting for more payload.
+        serverSendsInvalidContentLength(1, false);
+    }
+
+    @Test
+    public void serverSendsLargerContentLengthTrailers() throws Exception {
+        assumeTrue(h2PriorKnowledge); // http/1.x will timeout waiting for more payload.
+        serverSendsInvalidContentLength(1, true);
+    }
+
+    @Test
+    public void serverSendsSmallerContentLength() throws Exception {
+        serverSendsInvalidContentLength(-1, false);
+    }
+
+    @Test
+    public void serverSendsSmallerContentLengthTrailers() throws Exception {
+        serverSendsInvalidContentLength(-1, true);
+    }
+
+    private void serverSendsInvalidContentLength(int contentLengthAdder, boolean addTrailers) throws Exception {
+        InetSocketAddress serverAddress = bindHttpEchoServer(service -> new StreamingHttpServiceFilter(service) {
+            @Override
+            public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                        final StreamingHttpRequest request,
+                                                        final StreamingHttpResponseFactory responseFactory) {
+                return delegate().handle(ctx, request, responseFactory).flatMap(resp ->
+                        resp.toResponse().map(aggResp -> {
+                            aggResp.headers().remove(TRANSFER_ENCODING);
+                            aggResp.headers().set(CONTENT_LENGTH,
+                                    String.valueOf(aggResp.payloadBody().readableBytes() + contentLengthAdder));
+                            return aggResp.toStreamingResponse();
+                        }));
+            }
+        });
+        try (BlockingHttpClient client = forSingleAddress(HostAndPort.of(serverAddress))
+                .protocols(h2PriorKnowledge ? h2Default() : h1Default())
+                .executionStrategy(clientExecutionStrategy)
+                .buildBlocking()) {
+            HttpRequest request = client.get("/").payloadBody("a", textSerializer());
+            if (addTrailers) {
+                request.trailers().set("mytrailer", "myvalue");
+            }
+            if (h2PriorKnowledge) {
+                assertThat(invokeThrowableRunnable(() -> client.request(request)),
```

Comment:
not sure if there is a better way to assert multiple types of exceptions for junit4, and since junit5 PRs are coming in I'm hesitant to add `@Rules` for exception handling.

Reply:
this looks good, only one suggestion: you don't need `invokeThrowableRunnable` if you will use:
```java
Exception e = assertThrows(Exception.class, () -> client.request(request));
assertThat(e, either(...));
```

Reply:
good idea, updated